README.md
# Rawboot

## General Information

Rawboot is a general-purpose Discord bot intended to equip servers with a wide variety of functions including general chat moderation, music playback, current weather conditions and forecasts, and integration with popular game APIs to return player stats and other commonly sought information. As of this writing (8/11/2019) music and weather commands have been created along with some basic chat moderation tools, with the most immediate plans being the expansion of chat moderation capabilities including ban, kick, unban, and warnings.

Information on current commands and plans for improvement can be found further down in this readme and bot changes along with their dates can also be found in this repo in the changelog.md file. I have no current plans to create a website for this bot seeing as it's still a small-scale project, so this repo is the best place to stay up to date. The changelog file should be updated fairly often as new features and fixes are pushed.

## General Commands

- First mention goes to `!help`, which provides all the information in this section in an easy-to-read embed format. In the future this will also provide aliases and abbreviations for commands, e.g. `!vol` working in the same way that the full `!volume` command does.

- The `!ping` command pings the Discord server and relays the time elapsed with a standard "ping/pong" feedback format.

- The `!ascii <text>` command returns the input text in ASCII format. Capabilities are currently being expanded and will hopefully handle a wider variety of inputs in the future.

- The `!list` command lists all the custom emoji for the custom server in a small embed that is easy to read.

- The `!listnames` command returns a similar result to `!list` but also includes the name of each emoji. Each row is returned in `<emoji> | <emojiname>` format. This command can return a rather large result for servers with 50+ emoji, so it is recommended to use this sparingly and/or create a separate "Bot Commands" type channel to avoid cluttering chat servers.

- The `!join` command will make the bot join whichever channel the command issuer is currently in.

- The `!boot/!kick` command will make the bot leave the current voice channel. As with `!join`, this can only be used by someone in the voice channel.

- The `!delete <# of messages>` command deletes the specified number of previous messages from the current channel.

- The `!fellas` command has the bot join voice and play a short Hank Hill audio clip before leaving.

- The `!headpat` command will headpat the requester. Will add a way to use it on other people in the future.

- The `!counterstrike/!cs` command will play a popular League of Legends sound clip that likely needs no introduction.

### Weather

- The `!w <location>` command is used for current weather conditions at a given location. The bot accepts full city names, abbreviations, and ZIP codes to find location, e.g. `!w New York` and `!w ny` both return the weather for New York City and one of its many ZIP codes can be used for a slightly more specific scope.

- The `!forecast <location>` command provides a five-day weekday forecast for the current week at a given location. This command also accepts a variety of inputs to find location. At the time of this writing the API only returns weekday information, but I'm looking into ways to make weekend data available.

## Music Commands

### Please note that users must be in a voice channel to use music commands.

- The `!pause/!p` command pauses the current song playback. This is not a skip command so the queue is unaffected, just a simple pause button.

- The `!play <item to play>` is the command to initiate playback. This command accepts YouTube links (in either youtube.com or youtu.be formats) and can also play a search term. Note that the search term is the first result so specificity is helpful; `!play Debussy Arabesque No. 1` is considerably more likely to return the desired result than `!play arabeseque`.

- The `!queue/q` command shows a list of currently queued songs in the order they will be played, with the currently playing song highlighted at the top.

- The `!reboot` command reboots the bot in the event of any issues or changes. This command can only be used with sufficient server privileges, so most users will not have access to it.

- The `!reload <commandname>` command is similar to rebooting but is used to reload an individual command. This can be used by all server privilege levels and will likely fix any problems before an `!reboot` becomes necessary.

- The `!resume/!r` command is used to resume playback after it has been paused with `!pause/!p`. As with the pause command, this has no effect on queue position or playback order.

- The `!search <term>` command is used to return the top 5 YouTube search results for the provided term. This will provide both video titles and direct links, making usage with the `!play` command straightforward.

- The `!skip` command initiates a vote to skip the currently playing song. Users have a ten-second window in which to vote and the song is skipped if a majority of users in the voice channel vote to skip. In the event that the number of voice channel users is one or two, a single `!skip` is sufficient.

- The `!stats` command provides current information on the number of servers using the music queue. Returns the number of songs queued and the number of servers queueing songs.

- The `!stop` command is used to stop the entire queue. It is important to note that this not only stops the current song but clears all queue entries as well.

- The `!volume/vol <number>` command is used to set the playback volume of the current song to a percentage between 0 and 100. Default playback volume is 70 percent. The command will provide error messages for non-numerical feedback and numbers out of range, but decimal values (e.g. `!vol 55.7`) are allowed. The `!vol up/u` and `!vol down/d` commands will increase or decrease volume by ten percent respectively. Using `!vol` with no value provided will display the current volume.

## Game Commands

Will be added as they become available.

## Upcoming Plans and Improvements

1. The number one planned feature for right now is to add the ability to use the `!search` command results to play something directly from the list. Looking at something like returning each search result with a number and then a simple entry of the number should play the given result.

2. Looking to also add custom server prefix capabilities at some point to avoid any issues with other bots.

3. Expanding the capabilities of the `!ascii` command to format text a little better for larger entries and add image support.

4. Addition of a `!repeat` command to allow a song to be re-queued as easily as possible.

5. I'm looking at adding a `!loop` command that's similar to `!repeat`. The `!repeat` command will simply add the current song to the end of the queue again, whereas the `!loop` command will continually add the current song to the queue after a certain point. `!loop` will likely require higher privileges to prevent misuse and in most cases the function of `!repeat` will be more desirable anyway.

6. Add `!voteskip` as another option rather than regular skipping.

7. Provide a way for users (primarily other devs) to report potential security issues so they can be addressed.

## Feedback

Please feel free to contact me at evan.carlstrom@gmail.com with any issues or feedback!

changelog.md
# Changelog

**October 31, 2020**

- 🎃 A couple general package updates to improve playback stability. 🎃

**October 30**

- Patched some of the playback-related dependencies to hopefully fix those issues. Good results so far, testing locally for a couple days before relaunching live version.

**October 29, 2020**

- Bot offline for playback and other command troubleshooting.

**October 23, 2020**

- More minor updates and small security fixes.

**October 20, 2020**

- General package updates.

**October 1, 2020**

- As per requests from a couple people, currently waiting for Twitter API approval so that a simple crawler for tweets and retweets can be implemented.

**September 21, 2020**

- Working on making improvements to various forms of error handling to hopefully prevent issues that can occasionally arise.

**September 9, 2020**

- Moved to a new hosting service. Will monitor efficiency and persistence over the next week or so.

**September 3, 2020**

- Minor package updates to address potential security vulnerabilities.

**July 27, 2020**

- Patches a couple of DoS exploit vulnerabilities present in various packages.

**June 2, 2020**

- Playback issues have been solved!

**June 1, 2020**

- Testing a fix for recent playback errors, build seems stable so far and works correctly.

**May 29, 2020**

- Updates to playback dependencies to fix some of the errors in object data retrieval.

**May 23, 2020**

- Properly rounded values are now displayed after `!vol up` and `!vol down` are used. All other values should remain the same and display correctly.

**April 17, 2020**

- Updating some playback dependencies for more accurate Youtube data.

**April 10, 2020**

- Occasional playback issues with `!skip` and `!stop` seem to be fixed.

**April 9, 2020**

- A couple more ASCII art package updates for security.

**April 4, 2020**

- Potentially exploitable vulnerability in a dependency has been patched.

**April 3, 2020**

- General package upgrades and maintenance. Potential security vulnerabilities in a couple of dependencies have been patched.

**March 21, 2020**

- General package upgrades and maintenance.

**March 17, 2020**

- Only users in the same voice channel as the bot can add songs to the queue.

**March 10, 2020**

- General updates, League API access renewed for future work.

**Feburary 25, 2020**

- General package upgrades and maintenance.

**February 16, 2020**

- All ASCII art dependencies have been updated. This update also addresses some minor security concerns with those dependencies.

**February 11, 2020**

- `!vol up` and `!vol down` have been added. Still working out a minor issue with this but the command are working properly.

**January 24, 2020**

- General package upgrades.

**January 21, 2020**

- Latest small update has been pushed to the live version. This includes improved loading and reloading of commands, `!ascii` stability updates, and the addition of a couple of command aliases for convenience.

**January 19, 2019**

- Additional package updates that should help the stability of the `!ascii` command.

**January 18, 2020**

- Added the `!counterstrike` command as per server request.

**January 9, 2020**

- Bot loading and reloading times have been improved.

**January 3, 2020**

- New bot version released! Playback issues fixed and general improvements to stability have been made.

**January 2, 2020**

- All playback issues seem to have been fixed and YouTube audio is working correctly on all tests. The latest stable version of the bot with these changes will be released tomorrow.

**December 31, 2019**

- Happy new year! Nothing changing today.

**December 30,2019**

- Some more (hopefully final) package updates. Next released is scheduled for this Friday, 1/3/2020.

**December 24, 2019**

- Getting together some final changes for the next bot version that should hopefully come within the week. This update should fix the recent playback issues and make some other general improvements that will be detailed upon release.

**December 23, 2019**

- Continuing package updates. There are two newly introduced ones in particular that seem to be causing the issues, but I'm getting closer to a solution.

**December 20, 2019**

- Updated some packages and dependencies that seem to be causing the intermittent playback issues on certain videos.

**December 13, 2019**

- Currently working on implementing a command that will update the bot directly from its Git repo in the event automatic updates and/or reboots fail. This will only be available to the highest level users but should provide another update option should the other methods fail.

**November 30, 2019**

- Default volume now 70 percent.

**November 20, 2019**

- Default playback volume is now 75 percent. The recent bot volume overhaul from Discord caused playback to be rather quiet and this is a change to compensate.

**October 30, 2019**

- A few more minor package updates and cleanups to improve stability.

**October 28, 2019**

- Ability to add channel/server announcements will hopefully be coming within the week.

**October 26, 2019**

- Added the `!headpat` command as per server request.

- Updated some packages that will hopefully improve stability.

**October 24, 2019**

- Continuing to test some stability and reboot command fixes. General stability seems to have improved over the last two weeks, still ironing out some of the crash causes.

**October 14, 2019**

- Still testing the recent fixes to stability. Too early for anything conclusive but general performance since those changes were pushed has been promising.

**October 13, 2019**

- Small oversight fixed: the `!stop` command now requires the command user to be in the same voice channel as the bot.

**October 11, 2019**


league.js
////////// GENERAL DEPENDENCIES //////////

// const Discord = require('discord.js');
// const config = require('dotenv').config();
// const client = new Discord.Client();
// const token = process.env.DISCORD_BOT_SECRET;
// const lol_api = process.env.LOL_API_KEY;
// const server = require('./server.js');
// const moment = require('moment');

////////// LEAGUE DEPENDENCIES //////////

// const fs = require('fs');
// const gm = require('gm');
// const request = require('request');
// const urlencode = ('urlencode');
// const roundTo = require('round-to');

/////////////////////////////// ***** LEAGUE API ***** ///////////////////////////////

// League commands here, using the functions in the section below

/* possible commands:
!build
!game
!player
!runes? // will be different now with new system
!masteries?
!championinfo (tips, matchups, etc.)?
*/

// client.on('message', (message) => {
//   const member = message.member;
//   const msg = message.content.toLowerCase(); // easier to handle text inputs this way
//   const args = message.content.split(' ').slice(1).join(" ");
//   const argsTwo = message.content.split(' ').slice(1);
//   const channel = message.channel;
// });
// // placeholder - getPlayerID(), getMatch(), getChampionID(), etc. - add in builds?

// // will see how new runes work as opposed to getrunes/get masteries endpoints for the old setup

// let queues = {
//   '0': 'Custom',
//   '2': 'Normal 5v5 Blind Pick',
//   '14': 'Normal 5v5 Draft Pick',
//   '4': 'Ranked Solo 5v5',
//   '6': 'Ranked Premade 5v5', // does this ID still work with flex queues?
//   '65': 'ARAM'

//   // basic setup for now, there may be some overlap with new/old i.e. 'Ranked Solo 5v5' vs. 'Ranked Solo'
// };

// let maps = {
//   '1': `Summoner's Rift`,
//   '4': 'Twisted Treeline',
//   '12': 'Howling Abyss'

//   // there is also another SR map at ID 11 and TT at 10, not sure what any differences are
// };

// const liveMatch = "https://na1.api.riotgames.com/lol/spectator/v3/active-games/by-summoner/"; // constant path to retrieve live match data
// const playerID = "https://na1.api.riotgames.com/lol/summoner/v3/summoners/by-name/"; // constant path to retrieve summoner names (player names)
// const championID = "https://na1.api.riotgames.com/lol/static-data/v3/champions?locale=en_US&dataById=false&api_key=" + lol_api;
// const getChampion = "http://ddragon.leagueoflegends.com/cdn/6.24.1/data/en_US/champion.json";
// const routeInfo = 'http://api.champion.gg/v2/champions?limit=200&champData=hashes,firstitems,summoners,skills,finalitemshashfixed,masterieshash&api_key=' + champggToken;


// // ///// League API functions: getPlayerID(), getMatch(), getChampionID(), matchInfo(), championBuild()

// function getPlayerID(player, cb) {
    
//   request(playerID + urlencode(summoner) + "?api_key=" + lol_api, function(error, response,body) {
//     if(err) {
//       cb(err); // error handling
//     } else {
//       // parse the returned data and create a summoner object to contain it
//       let dataJSON = JSON.parse(response.body);
//       let summonerid = dataJSON.id;
//       let accountLevel = dataJSON.summonerLevel;
//       let profileID = dataJSON.profileIconId;
//       let summonername = dataJSON.name;
//       let summonerObject = {
//         "summonerid": summonerid,
//         "accountlvl": accountlvl,
//         "profileid": profileID,
//         "name": summonername
//       }
//       cb(false, summonerObject);
//     }
//   });
// } // end getPlayerID()

// function getMatch(summonerObject, cb) {

//   request(liveMatch + summonerObject.summonerid + "?api_key=" + lol_api, function(err, response, body) {
//     if(response.statusCode == 404) {
//       cb('The player is not currently in a live match.'); // simple error handling for likely the most frequent issue
//     } else {
//       requesterror(liveMatch, response.statusCode, function(err) {
//         if(err) {
//           cb(err); // generic error handler
//         } else {
//           // block occurs only when the given player is in a match
//           console.log('Parsing match data.');
          
//           // as with getPlayerID(), parse data and create an object. This is an object for a live game rather than a player.
//           let dataJSON = JSON.parse(response.body);
//           let gameID = dataJSON.gameId;
//           let gameMode = dataJSON.gameMode;
//           let mapID = dataJSON.mapID;
//           let gameType = dataJSON.gameType;
//           let gameStart = dataJSON.gameStartTime;
//           let players = dataJSON.participants;
//           let gameObject = {
//             "gameid": gameid,
//             "gamemode": gamemode,
//             "mapid": mapid,
//             "gametype": gameType,
//             "gametime": gametime,
//             "participants": participants,
//             "queue": dataJSON.gameQueueConfigId
//           }
//           cb(false, gameObject);
//         }
//       });
//     }
//   });
// } // end getMatch()

// function getChampionID(championName, cb) {

//   request(championID, function(error, response, body) {
//     requesterror(championID, response.statusCode, function(err) {
//     if(err) {
//       cb(err); // basic error handling
//     } else {
//       let dataJSON = JSON.parse(response.body);
//       // normalizing champion name input (capitalization, spacing, etc.) to minimize edge cases
//       // the next two lanes make every name lowercase and then capitalize the first letter
//       // this will hopefully cut down the number of elifs required to manage other cases

//       championname = championname.toLowerCase();
//       championname = championname.charAt(0).toUpperCase() + championname.slice(1);

//       // add cases for potential exceptions: MF, Kai'sa, Kog, ASol, J4, Lee, Xin, Yi, Mundo, TF, Rek'sai
//       // more: Wukong? (MonkeyKing in JSON, will investigate), Cho, Vel, Nunu? (will check with new name)
//       // can do an includes() for nicknames as well: Cass, Kass, Malph, etc.
        
//     // 3/13/20 note: grab names/info of all newer champion releases not on this list along with nicknames

//       // for the purpose of the following statements, the number "data" is assigned to references the individual champion keys on Riot's end
//       if(championname.includes("Miss") || championname.includes("Fortune")) {
//         data = 21;
//       } else if(championname.includes("Kai'")) {
//         // this is not in the Riot data I'm looking at, will add in
//       } else if(championname.includes("Kog")) {
//         data = 96;
//       } else if(championname.includes("Aurelion") || championname === "ASol") {
//         data = 136;
//       } else if(championname.includes("Jarvan" || championname === "J4")) {
//         data = 59;
//       } else if(championname === "Leesin" || championname === "Lee sin") { // using exact matches here because "Lee" will cause errors with Nidalee
//         data = 64;
//       } else if(championname.includes("Xin")) {
//         data = 5;
//       } else if(championname.includes("Master") || championname === "Yi") {
//         data = 11;
//       } else if(championname.includes("Mundo")) {
//         data = 36;
//       } else if(championname.includes("Twisted") || championname.includes("Fate") || championname === "TF") {
//         data = 4;
//       } else if(championname.includes("Rek")) {
//         data = 421;
//       } else if(championname.includes("Wu")) {
//         data = 62; // not spaced or a name with an apostrophe, but Wukong is internally known in Riot's database as "MonkeyKing" so this requires more specificity
//       } else if(championname.includes("Cho")) {
//         data = 31;
//       } else if(championname === "Velkoz" || championname === "Vel" || championname === "velkoz") { // using exact matches again because "vel" also appears in Evelynn's name
//         data = 161;
//       } else if(championname.includes("Nunu") || championname.includes("Willump")) { // probably not required but covering an extra base here since he's now named "Nunu & Willump"
//         data = 20;
//       }
//       // may add some more catches for nicknames in the future, just getting a standard working version for now
//         else {
//           try {
//             let data = dataJSON.data[championname].id;
//           } catch(err) {
//             cb('Not recognized as a valid champion name. Please try again!');
//             return;
//           }
//         }
//         cb(false, data);
//     }
//   });
//  });
// } // end getChampionID()

// // function matchInfo(matchObject, summonerObject, cb) {

// //   let placeholder = matchObject.queue;
// //   let gametype = queues[placeholder];
// //   // placeholder then gets the type of map that has been queued for:
// //   placeholder = matchObject.mapid;
// //   let map = mapname[placeholder];
// //   let matchid = matchObject.gameid;
// //   let time = matchObject.gametime;

// //   // time variables
// //   let ms = (new Date).getTime(); // milliseconds
// //   time = ((time - ms) * -1);
// //   let second = (time / 1000) % 70 - 0.5;
// //   let minute = (time / (1000 * 60)) % 60 - 0.5;
// //   let hour = (time / (1000 * 60 * 60)) - 0.5;
// //   // round all time variable values
// //   second = roundTo(second, 0);
// //   minute = roundTo(minute, 0);
// //   hour = roundTo(hour, 0);

// //   // two possible cases: match is live and there's a time or players are still loading in (no game time)
// //   if(hour > 100) {
// //     time = "Players loading!";
// //   } else {
// //     time = `Game live for ${hour} hours, ${minute} minutes, ${second} seconds.`;
// //     // may add another case here for anything below two minutes returning "one minute" instead of `${minute} minutes`, but this is a nitpicky detail.
// //   }

// //   // player-related variables and functions
// //   let players = matchObject.participants;
// //   let bluePlayers = [];
// //   let redPlayers = [];
// //   let team; // players will be moved from the empty objects based on their team

// //   // gathers player information and sorts them onto the appropriate teams
// //   for(let i = 0; i < players.length; i++) {
// //     let playerObject = {};
// //     if(players[i].teamId === 100) {
// //       playerObject.summonername = players[i].summonerName;
// //       playerObject.championId = players[i].championId;
// //       playerObject.summonerid = players[i].summonerId;
// //       playerObject.runes = players[i].runes;
// //       playerObject.masteries = players[i].masteries; // not sure how runes/masteries work with the new combined system, will see what's returned
// //       playerObject.team = "BLUE";
// //     }
// //     bluePlayers.push(playerObject);
// //     if(summonerObject.summonerid == players[i].summonerId) {
// //       team = 'BLUE';
// //     } else {
// //         playerObject.summonername = players[i].summonerName;
// //         playerObject.championId = players[i].championId;
// //         playerObject.summonerid = players[i].summonerId;
// //         playerObject.runes = players[i].runes;
// //         playerObject.masteries = players[i].masteries; // not sure how runes/masteries work with the new combined system, will see what's returned
// //         playerObject.team = "RED";

// //         redPlayers.push(playerObject);
// //         if(summonerObject.summonerid == players[i].summonerId) {
// //           team = 'RED';
// //       }
// //     }
// //   }

// //   // general match information
// //   matchObject = {
// //     'gametype': gametype,
// //     'map': map,
// //     'time': time,
// //     'blueplayers': blueplayers,
// //     'redplayers': redplayers,
// //     'team': team
// //   }
// //   liveMatchChampion(matchObject, function(err, newMatch) {
// //     if(err) {
// //       cb(err);
// //     } else {
// //       cb(false, newMatch, summonerObject);
// //     }
// //   });
// // } // end matchInfo()

// function liveMatchChampion() {

//   request(getChampion, function(error, response, body) {
//     requesterror(getChampion, response.statusCode, function(err) {
//       if(err) {
//         cb(err);
//       } else {
//         let dataJSON = JSON.parse(response.body);
//         let secondJSON = dataJSON.data;
//         for(let i = 0; i < matchObject.blueplayers.length; i++) {
//           for(let key in secondJSON) {
//             if(secondJSON[key].key == (matchObject.blueplayers[i].championid + "")) {
//               matchObject.blueplayers[i].championname = secondJSON[key].id;
//             // once again we handle exceptions, these being newer champions. Still have to find Kai'sa in the JSON and add her.
//             } else if(matchObject.blueplayers[i].championid === 141) {
//               matchObject.blueplayers[i].championname = "Kayn";
//             } else if(matchObject.blueplayers[i].championid === 498) {
//               matchObject.blueplayers[i].championname = "Xayah";
//             } else if(matchObject.blueplayers[i].championid === 497) {
//               matchObject.blueplayers[i].championname = "Rakan";
//             } else {
//               continue;
//             }
//           }
//         }
//         // now do the same exact thing, but loop through the red players instead.
//         for(let i = 0; i < matchObject.blueplayers.length; i++) {
//           for(let key in secondJSON) {
//             if(secondJSON[key].key == (matchObject.blueplayers[i].championid + "")) {
//               matchObject.blueplayers[i].championname = secondJSON[key].id;
//             // once again we handle exceptions, these being newer champions. Still have to find Kai'sa in the JSON and add her.
//             } else if(matchObject.blueplayers[i].championid === 141) {
//               matchObject.blueplayers[i].championname = "Kayn";
//             } else if(matchObject.blueplayers[i].championid === 498) {
//               matchObject.blueplayers[i].championname = "Xayah";
//             } else if(matchObject.blueplayers[i].championid === 497) {
//               matchObject.blueplayers[i].championname = "Rakan";
//             } else {
//               continue;
//             }
//           }
//         }
//       }

//       // leaving a placeholder here for future features.
//       // can handle runes, masteries, champion mastery, etc. in this function
//     });
//   });
// } // end liveMatchChampion()

// // // get build information for a specific champion
// function championBuild(champID, argsTwo, cb) {

//   console.log('Parsing build data.');
//   // handling Riot naming conventions here, the following line is required because their internal role data is uppercase
//   let role = argsTwo.slice(1).join('_').toUpperCase
//   // these lines handle the three roles that are called different things internally and externally
//   // jungle/top are the same but in uppercase
//   if(role == 'ADC' || role == 'AD' || role == 'DUOCARRY') {
//     role = 'DUO_CARRY';
//   } else if(role = 'SUPPORT' || role == 'SUPP' || role == 'DUO_SUPPORT') {
//     role = 'DUO_SUPPORT';
//   } else if(role == 'MID' || role == 'MIDLANE') {
//     role = 'MIDDLE';
//   }
//   request(routeInfo, function(err, response, body) {
//     requesterror(routeInfo, response.statusCode, function(err) {
//       if(err) {
//         cb(err); // basic error handler
//       } else {
//         let dataJSON = JSON.parse(response.body);
//         let champObject = {};
//         let notfound = true;

//         if(role == "") {
//           for(var key in dataJSON) {
//             if(dataJSON[key].championId == champid) {
//               champObject = dataJSON[key];
//               notfound = false;
//               break;
//             }
//           }
//         } else {
//           for(var key in dataJSON) {
//             if(dataJSON[key].championId == champid && dataJSON.role == role) {
//               champObject = dataJSON[key];
//               notfound = false;
//               break;
//             }
//           }
//         }
//       }
//     })
//   }
// } // end championBuild()